lte/gateway/deploy/roles/magma/files/update_mme_config_for_sanity.sh

```diff
@@ -0,0 +1,98 @@
+  # In case the backup file is already present, it means there was failure
+  # in last sanity run and current configuration file is already modified.
+  # Hence, this function will restore the same backup file before modifying
+  # it again, otherwise MME will crash in reading configuration from file
+  if [[ -f $mme_config_backup_file ]]; then
+    cp $mme_config_backup_file $mme_config_file
+  else
+    cp -n $mme_config_file $mme_config_backup_file
+  fi
+}
+
+function configure_multiple_plmn_tac {
+  # Remove default PLMN and TAC from MME configuration file
+  sed -i -e '/GUMMEI_LIST/{n;d}' -e '/TAI_LIST/{n;N;N;N;N;N;N;d}' \
+    -e '/TAC_LIST/{n;N;N;d}' $mme_config_file
+
+  # Configure multiple PLMNs and TACs in MME configuration file
+  gummei_config=(
+    '{ MCC: "001"; MNC: "01"; MME_GID: "1"; MME_CODE: "1" }'
+    '{ MCC: "001"; MNC: "02"; MME_GID: "1"; MME_CODE: "1" }'
+    '{ MCC: "001"; MNC: "03"; MME_GID: "1"; MME_CODE: "1" }'
+    '{ MCC: "001"; MNC: "04"; MME_GID: "1"; MME_CODE: "1" }'
+    '{ MCC: "001"; MNC: "05"; MME_GID: "1"; MME_CODE: "1" }'
+  )
+  gummei_cmd_str=""
+  for config in "${gummei_config[@]}"
+  do
+    gummei_cmd_str=$gummei_cmd_str"\ \ \ \ \ \ \ \ "$config",\n"
+  done
+  gummei_cmd_str=`echo $gummei_cmd_str | sed -e 's/\(.*\),\\\n/\1 /'`
+
+  tac_config=(
+    '{ MCC: "001"; MNC: "01"; TAC: "1" }'
+    '{ MCC: "001"; MNC: "02"; TAC: "2" }'
+    '{ MCC: "001"; MNC: "03"; TAC: "3" }'
+    '{ MCC: "001"; MNC: "04"; TAC: "4" }'
+    '{ MCC: "001"; MNC: "05"; TAC: "5" }'
+  )
+  tac_cmd_str=""
+  for config in "${tac_config[@]}"
+  do
+    tac_cmd_str=$tac_cmd_str"\ \ \ \ \ \ \ \ "$config",\n"
+  done
+  tac_cmd_str=`echo $tac_cmd_str | sed -e 's/\(.*\),\\\n/\1 /'`
+
+  sed -i -e "/GUMMEI_LIST/a $gummei_cmd_str" \
+    -e "/TAI_LIST/a $tac_cmd_str" \
+    -e "/TAC_LIST/a $tac_cmd_str" \
+    $mme_config_file
+}
+
+function reduce_mobile_reachability_timer_value {
+  # Reduce the mobile reachability timer to 1 minute, so that it can
+  # quickly be tested as part of Sanity. The current default value of
+  # Mobile Reachability Timer is 54 minutes
+  sed -i '/^        T3412/s/54/1/' $mme_config_file
+}
+
+function restore_mme_config {
+  # Restore the MME configuration from the backup configuration file and
+  # delete the backup configuration file, so that MME will use latest
+  # configuration file in next sanity runs
+  if [[ -f $mme_config_backup_file ]]; then
+    cp $mme_config_backup_file $mme_config_file
+    rm -f $mme_config_backup_file
+  else
+    exit $RETURN_BACKUP_MISSING
+  fi
+}
+
+if [[ $1 == "modify" ]]; then
+  # Modify the MME configuration file so that all the sanity test cases pass
+  if [[ ! -f $mme_config_file ]]; then
+    exit $RETURN_CONFIG_MISSING
+  fi
+  create_backup_or_restore_mme_config
+  configure_multiple_plmn_tac
+  # Commented for now because of crash in MME due to this change
+  # reduce_mobile_reachability_timer_value
```

Comment:
what is the reason for the crash? we cannot add these to the sanity with the default 54 minute timer...

Reply:
There are 2 test cases which need this timer to reduce: s1aptests/test_attach_detach_timer_expiry.py and s1aptests/test_attach_mobile_reachability_timer_expiry.py. The default value of timer is 54 minutes and with delta time one test case will end in around 54 minutes while the other will end in 116 minutes. To verify these 2 test cases as part of sanity we are changing this timer value to 1 minuted to quickly test these 2 test cases. 

The crash is not because of this change. It is there but it occurs at a very late time, whereas as part of this change it is quickly observed before even sanity could pass. The crash is found while handling the expiry of mobile reachability timer. The reason for the crash is observed as one of the dangling Mobile Reachability Timer which is created but  without stopping it, the timer id is overwritten with new timer. So, by the time this dangling timer is expired, the UE context was already cleared and MME crashes while accessing it. 

From the list of test cases in Sanity suite, a set of following 7 test cases were found causing creation of this unnecessary timer:
s1aptests/test_enb_partial_reset.py
s1aptests/test_enb_complete_reset.py
s1aptests/test_attach_detach_enb_rlf_initial_ue_msg.py
s1aptests/test_outoforder_attach_complete_ICSR.py
s1aptests/test_sctp_shutdown_after_multi_ue_attach.py
s1aptests/test_multi_enb_partial_reset.py
s1aptests/test_multi_enb_complete_reset.py